### ISSUE TITLE
Question: Can this problem be solved by CVXPY?

### ISSUE THREAD

**Issue by marciorpcoelho** [cvxgrp/cvxpy#879]
First of all, thank you very much for such an awesome library. CVXPY has saved my life a lot of times and it's a bless to work with! But I've run into a problem that so far can't solve.

I'm trying to maximize the total value of a group of cars. Each car has it's own value (along with other characteristics like color, version, etc) and given a few options to the users, i want to suggest the best group of cars that maximize their total value.

An example of the cars and their characteristics:

      Model_Code Color   Version   Score
    0         A   White    X       359.4879
    1         B   Black    X       425.1045
    2         E   Black    Y       742.0248
    3         C   Yellow   Z       1353.8500
    4         D   White    Y       807.1862

And aside from the typical restrictions of positive only solutions and overall total vehicles, this is the code I'm using:

    scores = dataset['Score'].values.tolist()
    
    selection = cp.Variable(dataset.shape[0], integer=True)
    
    order_size_restriction = cp.sum(selection) <= sel_order_size
    total_value = selection * scores
    
    problem = cp.Problem(cp.Maximize(total_value), [selection >= 0, selection <= 100, order_size_restriction])
    
    result = problem.solve(solver=cp.GLPK_MI, verbose=False)

And this works flawlessly, but, as expected, the solver only suggests one single model (C, with the highest score). So I tried to implement some variability in the solution, by, for example, forcing the solution to have cars with at least 2 different colors.

And here is where my problems started.
In order to apply this type of constraint, I first created the dummy variables for the colors of each car like so:

    ext_color_dummies = np.array(pd.get_dummies(dataset['Color']))

Which provided something like:

       [1 0 0,  
        0 1 0,  
        0 1 0,  
        0 0 1,
        1 0 0]

Now if I multiply the above matrix by the solution, which, for an order of 10 cars would be:

    [0 0 0 10 0]

The result would be:

    [0 0 10]

Which tells me that from the three available colors (white, black, yellow) I'm choosing only one (10 vehicles of color yellow).

Now, going straight to the point, how would I force the solution to have more than one selected color?
Here's what I tried:

    ext_color_dummies = np.array(pd.get_dummies(dataset['Colour']))
    ext_color_multiplication = selection * ext_color_dummies   
    ext_color_restriction = cp.length(ext_color_multiplication)

And then added ext_color_restriction to the constraints in the problem variable mentioned above like so:

    problem = cp.Problem(cp.Maximize(total_value), 
                         [selection >= 0, 
                          selection <= 100, 
                          order_size_restriction, 
                          ext_color_restriction_2 >= sel_ext_color_count])

Where sel_ext_color_count is the number of different colors to include in the solution.
But, as the title implies, this is when I started getting the dreaded error:

> DCPError: Problem does not follow DCP rules. Specifically: The
> following constraints are not DCP: 3.0 <= length(var337 * [[0. 0. 0.
> 1.] [0. 0. 0. 1.] [0. 0. 1. 0.] [0. 1. 0. 0.] [0. 0. 0. 1.] [0. 0. 1. 0.] [0. 1. 0. 0.] [0. 0. 0. 1.] [0. 1. 0. 0.] [0. 0. 0. 1.] [0. 0. 0. 1.] [0. 0. 1. 0.] [0. 0. 0. 1.] [0. 0. 0. 1.] [1. 0. 0. 0.] [0. 0. 1. 0.]]) , because the following subexpressions are not: |-- length(var337 * [[0. 0. 0. 1.] [0. 0. 0. 1.] [0. 0. 1. 0.] [0. 1. 0.
> 0.] [0. 0. 0. 1.] [0. 0. 1. 0.] [0. 1. 0. 0.] [0. 0. 0. 1.] [0. 1. 0. 0.] [0. 0. 0. 1.] [0. 0. 0. 1.] [0. 0. 1. 0.] [0. 0. 0. 1.] [0. 0. 0. 1.] [1. 0. 0. 0.] [0. 0. 1. 0.]])

Note: Error is from the actual code and not the example provided here, hence why the dummy variable matrix has different dimensions.

I tried a different, albeit more hacky solution which was:

    ext_color_restriction = cp.sum(ext_color_multiplication / ext_color_multiplication, axis=0, keepdims=True)

But alas, the problem would not go away:

> DCPError: Problem does not follow DCP rules. Specifically: The
> following constraints are not DCP: 3.0 <= Sum(var383 * [[0. 0. 0. 1.]
> [0. 0. 0. 1.] [0. 0. 1. 0.] [0. 1. 0. 0.] [0. 0. 0. 1.] [0. 0. 1. 0.]
> [0. 1. 0. 0.] [0. 0. 0. 1.] [0. 1. 0. 0.] [0. 0. 0. 1.] [0. 0. 0. 1.]
> [0. 0. 1. 0.] [0. 0. 0. 1.] [0. 0. 0. 1.] [1. 0. 0. 0.] [0. 0. 1. 0.]]
> / var383 * [[0. 0. 0. 1.] [0. 0. 0. 1.] [0. 0. 1. 0.] [0. 1. 0. 0.]
> [0. 0. 0. 1.] [0. 0. 1. 0.] [0. 1. 0. 0.] [0. 0. 0. 1.] [0. 1. 0. 0.]
> [0. 0. 0. 1.] [0. 0. 0. 1.] [0. 0. 1. 0.] [0. 0. 0. 1.] [0. 0. 0. 1.]
> [1. 0. 0. 0.] [0. 0. 1. 0.]], 0, True) , because the following
> subexpressions are not: |-- var383 * [[0. 0. 0. 1.] [0. 0. 0. 1.] [0.
> 0. 1. 0.] [0. 1. 0. 0.] [0. 0. 0. 1.] [0. 0. 1. 0.] [0. 1. 0. 0.] [0. 0. 0. 1.] [0. 1. 0. 0.] [0. 0. 0. 1.] [0. 0. 0. 1.] [0. 0. 1. 0.] [0. 0. 0. 1.] [0. 0. 0. 1.] [1. 0. 0. 0.] [0. 0. 1. 0.]] / var383 * [[0. 0. 0. 1.] [0. 0. 0. 1.] [0. 0. 1. 0.] [0. 1. 0. 0.] [0. 0. 0. 1.] [0. 0. 1. 0.] [0. 1. 0. 0.] [0. 0. 0. 1.] [0. 1. 0. 0.] [0. 0. 0. 1.] [0. 0. 0. 1.] [0. 0. 1. 0.] [0. 0. 0. 1.] [0. 0. 0. 1.] [1. 0. 0. 0.] [0. 0. 1. 0.]]

Finally I caved in, browsed a [few][1] [topics][2] as well as [this question][3] and eventually found the [expression rules for CVX][4].
And respectivelly hit a brick wall as I don't undertand why my first approach is giving DCPError (I believe I understand why the second is giving DCPError).

I already posted this question on Stackoverflow but alas, no answer yet, hence why I decided to come straight to the source.

So, In summary, my two questions are:

 - How to ensure variability in the solution?
 - Why is my first approach throwing DCPError and is there a way to
   solve it?

  [1]: https://stackoverflow.com/questions/44343243/constraints-do-not-follow-dcp-rules-in-cvxpy
  [2]: https://stackoverflow.com/questions/52285474/convex-optimization-problem-does-not-follow-dcp-rules
  [3]: https://github.com/cvxgrp/cvxpy/issues/76
  [4]: http://cvxr.com/cvx/doc/dcp.html#expressions

**Comment by rileyjmurray** [comment]
The atom ``ext_color_restriction = cp.length(ext_color_multiplication)`` is quasi-convex. The same is true for your later reformation, based on taking ratios of expressions. You must call ``prob.solve(qcp=True)`` in order to attempt such problems.

**Comment by marciorpcoelho** [comment]
First of all, thank your answer. I know this isn't the best place for such questions, but, I'm kind of desperate in finding a way to solve this and second, I didn't know of Stack Exchange (even though I'm an avid user of Stack Overflow.

Going to the point, I got another error:

> DQCPError: The problem is not DQCP

Is there any reformulation you can recommend or any other approach to the problem?

**Comment by rileyjmurray** [comment]
My suggestion of Stack Exchange was actually off-base, I was really just meaning that Stack Overflow was a better venue (you can add a "cvxpy" tag, if you'd like).

I'm not personally proficient with DQCP models. It could be that even with DQCP, you cannot model this problem with cvxpy. I wouldn't know without taking a closer look at your problem. Please post this question to Stack Overflow with a cvxpy tag. Try to follow the Stack Overflow guidelines when it comes to minimal examples, etc...

I'm changing the title of your question back to its original form, since there's now a chance that this is just misunderstanding a modeling issue. If you have reason to believe there is a bug in cvxpy (you might come to this conclusion after reading the DQCP documentation in detail; https://www.cvxpy.org/tutorial/dqcp/index.html), then feel free to create a new issue which explains a possible bug. I am closing this issue entirely, since there is no specific claim of a bug.

**Comment by marciorpcoelho** [comment]
Thank you for your time.
I had already posted a question on stack overflow but no answers so far,
hence why I came to github.

I'll try to read the documentation again and see If i can find a solution.

Best regards,
Márcio Coelho

**Comment by rileyjmurray** [comment]
@marciorpcoelho I spoke with @SteveDiamond and @akshayka about this. They believe that it's reasonable for recent additions to cvxpy to have more general user support on GitHub (at least for a limited time after said addition). Since disciplined quasi-convex programming is a recent addition to cvxpy, I'm re-opening this issue, and tagging it appropriately. I'll leave it to Akshay and Steven to provide the actual support here.

**Comment by marciorpcoelho** [comment]
Thank you so much for your availability! 
If there's more information you need, please tell me.

**Comment by SteveDiamond** [comment]
@marciorpcoelho sorry for the slow response. I've been traveling. I don't know how best to address this problem. Perhaps you could add ``gamma*sum_squares(selection)`` as a regularization term, for an appropriate value of ``gamma``?

**Comment by marciorpcoelho** [comment]
Thank you for your response @SteveDiamond, I had never thought of adding a regularization term. Unfortunately, after playing around with this idea I wasn't able to reach a solution. Not only was i not able to find a suitable value for `gamma` (I have no sort of penalization by choosing only a single color) and using sum_squares(selection) doesn't necessarily enforce color distinction and I lose some information as `sum_squares([2 1 0])` is the same of `sum_squares([0 1 2])`.

**Comment by rileyjmurray** [comment]
Since this issue has been stagnant for a while, I'm going to close it. @marciorpcoelho if you think you encounter a cvxpy bug in the future, we absolutely welcome creating a new issue and investigating.
